### ISSUE TITLE
`SimpleImputer` should perform strategy-dependent data type inference

### ISSUE THREAD

**Issue by victornoel** [jpmml/sklearn2pmml#295]
Hi,

I'm converting a fitted sklearn pipeline that looks like this:
```python
categorical_transformer = sklearn2pmml.PMMLPipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinalencoder', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', categorical_transformer, categorical_cols),
        ('num', 'passthrough', numerical_cols),
    ],
    remainder = 'drop',
    verbose = True
)
pipeline = sklearn2pmml.PMMLPipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', xgb)
])
```

But in the resulting PMML, I can see this kind of thing in the `PMML.DataDictionnary` element:
```xml
<DataField name="Colour" optype="categorical" dataType="double">
  <Value value="Green"/>
  <Value value="Blue"/>
</DataField>
```

And so when trying to evaluate it using jpmml_evaluator, the evaluator tries to convert it to a number even though it shouldn't and we get a NumberFormatException.

Note that of course, the transformer is properly generated later in the PMML:
```xml
<DerivedField name="encoder(Colour)" optype="categorical" dataType="double">
```

**Comment by vruusmann** [comment]
As a workaround, start your pipeline with a `sklearn2pmml.decoration.ContinuousDomain` or `sklearn2pmml.decoration.CategoricalDomain` step, and the data type will be correct.

**Comment by victornoel** [comment]
very nice, thank you for the workaround :)

**Comment by victornoel** [comment]
@vruusmann unfortunately it doesn't seem to work as expected (or I misused it ^^), I changed my categorical_transformer pipeline like so:
```py
categorical_transformer = sklearn2pmml.PMMLPipeline(steps=[
    ('cat', CategoricalDomain()),
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinalencoder', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])
```

And now I'm getting this error during `fit`:
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/usr/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/tmpd0jsezj1", line 287, in <module>
  File "/usr/local/lib/python3.8/dist-packages/sklearn/compose/_column_transformer.py", line 470, in fit
    self.fit_transform(X, y=y)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/compose/_column_transformer.py", line 507, in fit_transform
    result = self._fit_transform(X, y, _fit_transform_one)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/compose/_column_transformer.py", line 434, in _fit_transform
    return Parallel(n_jobs=self.n_jobs)(
  File "/usr/local/lib/python3.8/dist-packages/joblib/parallel.py", line 1041, in __call__
    if self.dispatch_one_batch(iterator):
  File "/usr/local/lib/python3.8/dist-packages/joblib/parallel.py", line 859, in dispatch_one_batch
    self._dispatch(tasks)
  File "/usr/local/lib/python3.8/dist-packages/joblib/parallel.py", line 777, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "/usr/local/lib/python3.8/dist-packages/joblib/_parallel_backends.py", line 208, in apply_async
    result = ImmediateResult(func)
  File "/usr/local/lib/python3.8/dist-packages/joblib/_parallel_backends.py", line 572, in __init__
    self.results = batch()
  File "/usr/local/lib/python3.8/dist-packages/joblib/parallel.py", line 262, in __call__
    return [func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/joblib/parallel.py", line 262, in <listcomp>
    return [func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/utils/fixes.py", line 222, in __call__
    return self.function(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/pipeline.py", line 754, in _fit_transform_one
    res = transformer.fit_transform(X, y, **fit_params)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/pipeline.py", line 378, in fit_transform
    Xt = self._fit(X, y, **fit_params_steps)
  File "/usr/local/lib/python3.8/dist-packages/sklearn2pmml/pipeline/__init__.py", line 70, in _fit
    return super(PMMLPipeline, self)._fit(X = X, y = y, **fit_params)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/pipeline.py", line 303, in _fit
    X, fitted_transformer = fit_transform_one_cached(
  File "/usr/local/lib/python3.8/dist-packages/joblib/memory.py", line 352, in __call__
    return self.func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/pipeline.py", line 754, in _fit_transform_one
    res = transformer.fit_transform(X, y, **fit_params)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/base.py", line 699, in fit_transform
    return self.fit(X, **fit_params).transform(X)
  File "/usr/local/lib/python3.8/dist-packages/sklearn2pmml/decoration/__init__.py", line 139, in fit
    X = column_or_1d(X, warn = True)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py", line 921, in column_or_1d
    raise ValueError(
ValueError: y should be a 1d array, got an array of shape (5514099, 12) instead.
```

**Comment by vruusmann** [comment]
> ValueError: y should be a 1d array, got an array of shape (5514099, 12) instead.

The `CategoricalDomain` decorator works with single columns only.

The fact that your're processing all your categorical columns in one go also explains why data type detection fails. There's a good chance that things would works much better if you processed categorical columns one-by-one.

**Comment by victornoel** [comment]
> The CategoricalDomain decorator works with single columns only.

Out of curiosity, is that a limitation of the decorator due to its nature? I would have expected to be able to use like any other transformer in a pipeline, regardless of the number of columns, I wonder if this is an improvement that can be implemented to ease the life of users?

> The fact that your're processing all your categorical columns in one go also explains why data type detection fails. There's a good chance that things would work much better if you processed categorical columns one-by-one.

Ok, I didn't understand there was some "detection" happening, I just thought it was inferred from the transformers present, but I realize I was naive since SimpleImputer says nothing about the type of data :)

Would you say I was not hitting a bug but just a misconfigured pipeline? If yes I believe we can close this, except if there are obvious improvement to be done to help users.

**Comment by vruusmann** [comment]
>>  The CategoricalDomain decorator works with single columns only.
>
> Out of curiosity, is that a limitation of the decorator due to its nature?

It's due to my limited Numpy/Scikit-Learn/Python coding abilities.

Once you get multi-column support, you'd want to do _mixed-data type_ multi columns, etc.

Fundamentally, a decorator should work on one column ("give me all the specifics about this column, and this column only"). The fact that some decorators provide multi-column support (eg. `ContinuousDomain`) is an exception, not the rule.

> I didn't understand there was some "detection" happening, I just thought it was inferred from the transformers present, but I realize I was naive since SimpleImputer says nothing about the type of data :)

For Scikit-Learn everything is `float64`.

However, the real world is richer than that, hence the need for decoration.

> Would you say I was not hitting a bug but just a misconfigured pipeline?

The Python side error message definitely should be clearer/more informative. Right now it complains about the `y` variable, which is super-misleading.

SkLearn2PMML uses Scikit-Learn utility function for enforcing input array dimensions. Perhaps it should be replaced with a home-brewn one.

This Scikit-Learn utility function is also generating tons of warnings.

> If yes I believe we can close this, except if there are obvious improvement to be done to help users.

Nope, let's keep this issue open.

If the `SimpleImputer` uses `strategy=most_frequent` (as is the case here), then the column type detection logic should return `categorical+string` not `continuous+double`.

**Comment by victornoel** [comment]
> If the `SimpleImputer` uses `strategy=most_frequent` (as is the case here), then the column type detection logic should return `categorical+string` not `continuous+double`.

Actually, such an imputer could be used for floats for example: NAN would be replaced by the most frequent value. So unfortunately, I suppose the decorator is needed (or the detection happens correctly from the get go, I haven't looked at how the type detection is implemented)

**Comment by vruusmann** [comment]
> Such an imputer could be used for floats for example: NAN would be replaced by the most frequent value.

If the training dataset contains missing values, then it is likely that `SimpleImputer` transformation is the first step in the pipeline (I'd suggest starting the pipeline always with a decorator step, but that's that).

The point is that `strategy=most_frequent` is a special strategy case, and it's note correct to return `continuous+double` always. The correct data type can be inferred from the `SimpleImputer.statistic` attribute.

**Comment by vruusmann** [comment]
> If the training dataset contains missing values, then it is likely that SimpleImputer transformation is the first step in the pipeline (I'd suggest starting the pipeline always with a decorator step, but that's that).

TLDR: "Likely first step classes need better PMML conversion support"

**Comment by victornoel** [comment]
oh right, sorry I misunderstood what you were saying :) this would be great indeed. And also I agree and understand what you mean about using decorators all the time.

The only strange thing about using `CategoricalDomain` for me is that it seems to have some kind of overlapping with `SimpleImputer` in terms of features, e.g., missing value processing. I suspect that if I don't specify anything except `dtype` for `CategoricalDomain`, they won't conflict with each other (e.g., there won't be any missing values processing done), but I'm still uncomfortable when using it because of the doubts I have about this overlapping :)

**Comment by vruusmann** [comment]
I can't reproduce the reported behaviour locally - when I start the pipeline with a `SimpleImputer(strategy = "most_frequent")` step and pass multiple string columns to it, then the field type is correctly detected as `categorical+string` (and not `categorical+double` as reported here).

> categorical_transformer = sklearn2pmml.PMMLPipeline(..)

The `sklearn2pmml.pipeline.PMMLPIpeline` class should be used as the parent/topmost-level step container only. The SkLearn2PMML/JPMML-SkLearn stack should issue a warning if you're using it at nested levels; if you don't see that warning, then perhaps you're using a badly outdated SkLearn2PMML package version (that might also explain the `SimpleImputer` behaviour).

**Comment by victornoel** [comment]
@vruusmann thanks for looking into it. 

> I can't reproduce the reported behaviour locally

Yeah, I suspect the data we were manipulating with that pipeline wasn't correctly cleaned and some `np.nan` were still present even though all other values were strings. This must have influenced the detection of the type in the wrong direction.

For the record, we found a way to make it work by:
- ensuring we only have `None` or `str` in the data
- using `CategoricalDomain`, forcing the `dtype` parameter to `str`, for each of the concerned columns
- use `SimpleImputer` to replace `None` with the `most_frequent` value

Now, it's possible that if we remove the `CategoricalDomain`, it will still work, but actually, you were correct that it's best to always use such a first transformer in order also to specify what should happen for unknown values.

And we are using the latest version yes. It's possible there were warnings that we missed :)
